### PR TITLE
Combinatorial_map: add save/load functionality in the LCC demo

### PR DESCRIPTION
**Pull request by gdamiand**
Improvement of the LCC demo.
Add save/load functions, a function to merge coplanar faces, update the OpenGL drawing to allow to draw non convex faces.

Tested in https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-60.shtml

### CODE REVIEW

File: Linear_cell_complex/demo/Linear_cell_complex/CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ endif()
 
 ## For profilling with gprof
 # add_definitions("-pg")
+add_definitions("-Wall -Wextra")
```
Comment (lrineau) [inline Linear_cell_complex/demo/Linear_cell_complex/CMakeLists.txt]: That cannot be accepted into CGAL: you cannot decide which compiler flags the users may want. What is more, unless I read wrongly, you enable flags on all platforms, and that will creates compilation errors on all platforms that do not understand those flags (old versions of gcc, and all Microsoft compilers).

File: Linear_cell_complex/demo/Linear_cell_complex/Viewer.h [hunk 2]
```diff
@@ -42,20 +51,34 @@ class Viewer : public QGLViewer
...
+      edges(true), vertices(true), m_displayListCreated(false),
+      m_previous_scene_empty(true)
   {
     QGLFormat newFormat = this->format();
     newFormat.setSampleBuffers(true);
     newFormat.setSamples(16);
     this->setFormat(newFormat);
+
+    FTess = gluNewTess();
+    gluTessCallback(FTess, GLU_TESS_BEGIN,(void (*) ( ))  & glBegin);
```
Comment (lrineau) [inline Linear_cell_complex/demo/Linear_cell_complex/Viewer.h]: Just for information, those lines use GLU, that is deprecated in recent versions of OpenGL, and do not exist for very recent platforms (such as Android, iOS).
My intern student Maxime (@maxGimeno) is working on renewing the OpenGL code of all 3D CGAL demos, and we have to replace the calls to GLU fonctions by something else.

File: Linear_cell_complex/demo/Linear_cell_complex/CMakeLists.txt (hunk 1 above)
Comment (gdamiand) [inline Linear_cell_complex/demo/Linear_cell_complex/CMakeLists.txt]: Removed (these flags were pushed by error).

File: Linear_cell_complex/demo/Linear_cell_complex/Viewer.h (hunk 2 above)
Comment (gdamiand) [inline Linear_cell_complex/demo/Linear_cell_complex/Viewer.h]: I agree, these code need to be modified when moving to new OpenGL code.
Note that in some application, I really need to draw non convex faces.

**Comment by lrineau** [comment]
The test results for CGAL-4.7-Ic-66 [are not good at the line of the LCC demo](https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-66.shtml#Linear_cell_complex_Demo).

**Comment by lrineau** [comment]
The results for CGAL-4.7-Ic-67 [are still very bad for the LCC demo](https://cgal.geometryfactory.com/CGAL/Members/testsuite/results-4.7-Ic-67.shtml##Linear_cell_complex_Demo).

What is more, the use of GLU and Glut is deprecated be recent versions of OpenGL. Maybe you should explain what you are doing with GLU/Glut, and GeometryFactory could help you use an alternative.

**Comment by gdamiand** [comment]
I need to draw non convex faces, this is the reason why I use glu tesselator.

I know glu and glut is deprecated, but this is a temporary patch before the switch to the new version of qt/opengl. Do you want I remove this code ?

**Comment by lrineau** [comment]
In the Polyhedron, we use code to triangulation non-convex faces of poyhedra or surface mesh, using only CGAL code: a `Constrained_Delaunay_triangulation_2`, with a special traits class that computes the 2D triangulation of 3D points in a plane. Could it be used by your demo too?

**Comment by gdamiand** [comment]
Sure a similar solution could be use.

But I don't think it is a good idea to work on this part before the future new version of qt/openGL.

This is why I propose to use glu tesselator...

**Comment by lrineau** [comment]
Well, Maxime Gimeno (@maxGimeno) is working on a branch to use OpenGL ES 2 (on Android) and OpenGL 3 core profile (for the desktop on Windows/Linux/MacOS). We expect to merge that in master this week or the next one. This is happening now. That is why  I am talking about the subject.

I do not write that to block your branch/pull-request. I just want to offer a collaboration to fix the compilation issues you have, and at the same time avoid using a deprecated feature of OpenGL, that will be removed from our use of OpenGL in CGAL next week.

**Comment by gdamiand** [comment]
I understand. If the merge will be in 2 or 3 weeks, I propose to remove my modification of drawing from this pull request (while keeping other modifications in the demo); then work after the merge of the work of Maxime on the drawing of non convex faces with the new version.

**Comment by lrineau** [comment]
Good. I agree with the plan.

Modification to demos can be accepted quite late in the release cycle.

**Comment by gdamiand** [comment]
Done.
Tell me when I can test the branch in integration (after the code freeze).

**Comment by lrineau** [comment]
Please update this pull-request by pushing the updated branch to `mine`.

See ["Update a pull-request" on the wiki](
https://cgal.geometryfactory.com/CGAL/Members/wiki/Guidelines/Source_Code_Management_GIT#Update_a_pull-request).

**Comment by gdamiand** [comment]
Done.

**Comment by lrineau** [comment]
No, you did not push to your fork. This pull-request has not been updated.

**Comment by gdamiand** [comment]
Here the copy of my terminal:

<gdamiand - ~/sources/cgal>  git push mine Combinatorial_map_save_load-gdamiand
Décompte des objets: 7, fait.
Delta compression using up to 8 threads.
Compression des objets: 100% (6/6), fait.
Écriture des objets: 100% (7/7), 1.58 KiB | 0 bytes/s, fait.
Total 7 (delta 4), reused 0 (delta 0)
To git@github.com:gdamiand/cgal.git
   cf563fa..fbfd478  Combinatorial_map_save_load-gdamiand -> Combinatorial_map_save_load-gdamiand

**Comment by lrineau** [comment]
Sorry, I got confused because in the pull-request the commits and the comments are ordered by date/time. But for a commit that is the date/time of the commit, not of the push. I can see the HEAD (fbfd478) of your branch in this pull-request.

**Comment by lrineau** [comment]
As expected, there is a big set of conflicts between this pull-request and the pull-request #154. :-(

**Comment by gdamiand** [comment]
Yes, this is normal, and this is probably the case for all the demo modified after the creation of the qt5 branch...

**Comment by lrineau** [comment]
The new Qt code is:
  - in conflict with the Qt5 branch,
  - different enough that I do not know how to deal with the conflicts.

I will be on vacation next week. I transfer the assignment to Sébastien, in case you can work together and with Maxime (@maxGimeno) to adapt the new code to Qt5.

**Comment by lrineau** [comment]
@gdamiand : Guillaume, what should we do with that pull-request? I have asked for help 11 days ago, and there was no answer. The issue is odd: we have important conflicts on the Qt4/Qt5 side. On the other hand, your branch was not supposed to modify the Qt4 code. That is why I am confused.

I am not opposed at all to the modification, but those technical issues are really painful. Do you have a few hours so that you can fork again against `master`, and re-create a brand-new branch that just adds the save/load functions?

**Comment by gdamiand** [comment]
I will do this modification, and probably by doing what you suggest (recreate a branch from master and integrate my modifications).
But I am to be sure that the qt5 demo is stable, in order to do not these modifications several times. And I am not sure this is the case (I have a segfault with the demo, cf my email on cgal-dev).

**Comment by lrineau** [comment]
Guillaume, now that the Qt5 version of CGAL demos work on your machine, can we try to resurrect that pull-request?

**Comment by gdamiand** [comment]
I was not sure that the branch containing the correction was integrated in master. Is it the case ?
If yes sure I will work on that.

**Comment by lrineau** [comment]
You are right. Some modifications are still in cgal-public-dev/CGAL-Qt5_support-GF. It will be tested again in `integration` tonight. I hope to be able to merge it in `master` tomorrow.

**Comment by gdamiand** [comment]
@Laurent I reintegrated the new code from a proper copy of master; I also improved the new drawing code; and I pushed it in Integration for tonight.

**Comment by lrineau** [comment]
Good. If the tests are good tonight, it will be in CGAL-4.7-beta2.

**Comment by lrineau** [comment]
Hi Guillaume (@gdamiand), you can now remove the branch `Combinatorial_map_save_load-gdamiand` from your fork:
```shell
git push mine --delete Combinatorial_map_save_load-gdamiand
```

**Pull request merged**